ProjFS.Mac/PrjFSKextTests/VnodeCacheTests.mm

```diff
@@ -0,0 +1,160 @@
+    VnodeCacheEntry* emptyArray = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
+    XCTAssertTrue(0 != memcmp(emptyArray, s_entries, sizeof(VnodeCacheEntry) * s_entriesCapacity));
+    
+    InvalidateCache_ExclusiveLocked();
+    XCTAssertTrue(0 == memcmp(emptyArray, s_entries, sizeof(VnodeCacheEntry)*s_entriesCapacity));
+    
+    free(emptyArray);
+    FreeCacheEntries();
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_ReturnsStartingIndexWhenNull {
+    AllocateCacheEntries(/* capacity*/ 100, /* fillCache*/ false);
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertTrue(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    XCTAssertTrue(cacheIndex == startingIndex);
+    
+    FreeCacheEntries();
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_ReturnsFalseWhenCacheFull {
+    AllocateCacheEntries(/* capacity*/ 100, /* fillCache*/ true);
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertFalse(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    
+    FreeCacheEntries();
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_WrapsToBeginningWhenResolvingCollisions {
+    AllocateCacheEntries(/* capacity*/ 100, /* fillCache*/ true);
+    
+    uintptr_t emptyIndex = 2;
+    MarkEntryAsFree(emptyIndex);
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertTrue(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    XCTAssertTrue(emptyIndex == cacheIndex);
+    
+    FreeCacheEntries();
+}
+
+- (void)testTryFindVnodeIndex_SharedLocked_ReturnsLastIndexWhenEmptyAndResolvingCollisions {
+    AllocateCacheEntries(/* capacity*/ 100, /* fillCache*/ true);
+    uintptr_t emptyIndex = s_entriesCapacity - 1;
+    MarkEntryAsFree(emptyIndex);
+    
+    vnode_t testVnode = &TestVnode;
+    uintptr_t startingIndex = 5;
+    uintptr_t cacheIndex;
+    XCTAssertTrue(TryFindVnodeIndex_Locked(testVnode, startingIndex, /* out */ cacheIndex));
+    XCTAssertTrue(emptyIndex == cacheIndex);
+    
+    FreeCacheEntries();
+}
+
+- (void)testTryInsertOrUpdateEntry_ExclusiveLocked_ReturnsFalseWhenFull {
+    AllocateCacheEntries(/* capacity*/ 100, /* fillCache*/ true);
+
+    uintptr_t indexFromHash = 5;
+    vnode_t testVnode = &TestVnode;
+    uint32_t testVnodeVid = 7;
+
+    XCTAssertFalse(
+        TryInsertOrUpdateEntry_ExclusiveLocked(
+            testVnode,
+            indexFromHash,
+            testVnodeVid,
+            true, // invalidateEntry
+            TestRootHandle));
+
+    FreeCacheEntries();
+}
+
+static void AllocateCacheEntries(uint32_t capacity, bool fillCache)
+{
+    s_entriesCapacity = capacity;
+    s_entries = static_cast<VnodeCacheEntry*>(calloc(s_entriesCapacity, sizeof(VnodeCacheEntry)));
```

Comment:
The idiomatic C++ way to do this kind of alloc is with operator `new[]`, and as we're not in the kernel here, I see no reason to revert to C, which is what forces the cast. (The matching release needs to be with `delete[]` of course.)